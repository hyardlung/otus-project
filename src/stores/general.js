import { ref } from "vue";
import { defineStore } from "pinia";

export const useGeneralStore = defineStore("general", () => {
  const loader = ref("src/assets/ball-triangle.svg");
  const isLoading = ref(false);

  function getImageSrc(obj) {
    const src = obj.formats["image/jpeg"];
    return src ? src : "";
  }

  function strToArr(str) {
    return str.includes(",") ? str.split(", ") : [str];
  }

  return { loader, isLoading, getImageSrc, strToArr };
});
