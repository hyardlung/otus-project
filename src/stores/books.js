import axios from "axios";
import { ref, reactive } from "vue";
import { defineStore } from "pinia";
import { useGeneralStore } from "./general";
import { strToArr } from "@/helpers";

export const useBooksStore = defineStore("books", () => {
  const generalStore = useGeneralStore();

  const books = reactive([]);
  const currentBook = ref({});
  const currentPage = ref(1);
  const baseURL = ref("https://gutendex.com/books/");

  function getNumberOfBooks() {
    return books.length;
  }

  async function getBooksFromApi() {
    generalStore.isLoading = true;
    const response = await axios.get(baseURL.value);
    books.value = response.data.results;
    generalStore.isLoading = false;
  }

  async function prevPage() {
    generalStore.isLoading = true;
    if (currentPage.value > 1) {
      const response = await axios.get(
        `${baseURL.value}?page=${currentPage.value--}`
      );
      books.value = response.data.results;
    }
    generalStore.isLoading = false;
  }

  async function nextPage() {
    generalStore.isLoading = true;
    const response = await axios.get(
      `${baseURL.value}?page=${currentPage.value++}`
    );
    books.value = response.data.results;
    generalStore.isLoading = false;
  }

  function findBook(id) {
    currentBook.value = books.value.find((item) => item.id == id);
  }

  function addBook(book, coverUrl, authors) {
    book.id = Math.random();
    book.formats = {
      "image/jpeg": coverUrl.value || "./src/assets/book-cover-placeholder.svg",
    };
    book.authors = [];
    book.authors.push(...authors);
    book.subjects = strToArr(book.subjects);
    book.bookshalves = strToArr(book.bookshalves);
    books.value.push(book);
  }

  function deleteBook(idx) {
    books.value.splice(idx, 1);
  }

  return {
    books,
    currentPage,
    currentBook,
    getNumberOfBooks,
    getBooksFromApi,
    findBook,
    addBook,
    deleteBook,
    prevPage,
    nextPage,
  };
});
