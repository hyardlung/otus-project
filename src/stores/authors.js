import { ref, reactive } from "vue";
import { defineStore } from "pinia";
import { useBooksStore } from "./books";

export const useAuthorsStore = defineStore("authors", () => {
  const booksStore = useBooksStore();

  const authors = reactive([]);
  const newAuthor = ref({});
  const currentAuthor = ref({});

  async function createAuthorsArray() {
    if (!booksStore.books.value) await booksStore.getBooksFromApi();
    await booksStore.books.value.forEach((book) => {
      book.authors.forEach((author) => {
        authors.push(author);
      });
    });
  }

  function findAuthor(name) {
    currentAuthor.value = authors.find((author) => author.name === name);
  }

  function AddAuthor(author) {
    authors.push(author);
    newAuthor.value = {};
  }

  return {
    authors,
    newAuthor,
    currentAuthor,
    createAuthorsArray,
    findAuthor,
    AddAuthor,
  };
});
