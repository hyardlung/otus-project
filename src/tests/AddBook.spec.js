import { describe, expect, it, vi, beforeEach, afterEach } from "vitest";
import { mount } from "@vue/test-utils";
import { setActivePinia, createPinia } from "pinia";
import { createTestingPinia } from "@pinia/testing";
import { useBooksStore } from "@/stores/books";

import component from "../views/AddBook.vue";

describe("AddBook", () => {
  let wrapper;
  let booksStore = null;
  setActivePinia(createPinia());
  booksStore = useBooksStore();

  beforeEach(() => {
    wrapper = mount(component, {
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
          }),
        ],
        mocks: [booksStore],
        stubs: ["RouterLink"],
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("mounts without errors", () => {
    expect(wrapper).toBeDefined();
    expect(wrapper.exists()).toBe(true);
  });

  it("renders all", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render form inputs and button", () => {
    // Assert form inputs
    expect(wrapper.find('[data-test="book-title-input"]').exists()).toBe(true);
    expect(wrapper.find('[data-test="book-cover-input"]').exists()).toBe(true);
    expect(wrapper.find('[data-test="subjects-input"]').exists()).toBe(true);
    expect(wrapper.find('[data-test="bookshalves-input"]').exists()).toBe(true);

    // Assert button
    expect(wrapper.find(".add-form__button").exists()).toBe(true);
  });

  it('should add new author when clicking "Add another author" button', async () => {
    const nameInput = wrapper.find('[data-test="author-name-input"]');
    const birthInput = wrapper.find('[data-test="author-birth-input"]');
    const addAuthorBtn = wrapper.find(".add-form__add-author");

    const addAuthorToNewBook = vi.fn(() => []);

    // Fill in author fields
    await nameInput.setValue("John Doe");
    await birthInput.setValue("1990");

    // Click "Add another author" button
    await addAuthorBtn.trigger("click");
    addAuthorToNewBook.mockReturnValue([
      { name: "John Doe", birth_year: "1990" },
    ]);

    expect(addAuthorToNewBook()).toStrictEqual([
      { name: "John Doe", birth_year: "1990" },
    ]);

    // Fill in second author fields
    await nameInput.setValue("Agent Smith");
    await birthInput.setValue("1968");
    await addAuthorBtn.trigger("click");
    addAuthorToNewBook.mockReturnValue([
      { name: "John Doe", birth_year: "1990" },
      { name: "Agent Smith", birth_year: "1968" },
    ]);

    expect(addAuthorToNewBook()).toStrictEqual([
      { name: "John Doe", birth_year: "1990" },
      { name: "Agent Smith", birth_year: "1968" },
    ]);
  });
});
