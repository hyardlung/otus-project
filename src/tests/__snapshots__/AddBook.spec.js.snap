// Vitest Snapshot v1

exports[`AddBook > renders all 1`] = `
"<main data-v-271f3949=\\"\\">
  <div class=\\"container\\" data-v-271f3949=\\"\\">
    <div class=\\"controls\\" data-v-271f3949=\\"\\">
      <router-link-stub to=\\"[object Object]\\" data-v-271f3949=\\"\\"></router-link-stub>
    </div>
    <form class=\\"add-form\\" data-v-271f3949=\\"\\">
      <h1 data-v-271f3949=\\"\\">Add new book</h1><!-- book title --><label class=\\"add-form__item\\" data-v-271f3949=\\"\\"> Title* <input type=\\"text\\" placeholder=\\"Book title\\" data-test=\\"book-title-input\\" class=\\"add-form__input\\" data-v-271f3949=\\"\\"></label><!-- book cover --><label class=\\"add-form__item\\" data-v-271f3949=\\"\\"> Cover <input type=\\"text\\" placeholder=\\"Cover URL\\" data-test=\\"book-cover-input\\" class=\\"add-form__input\\" data-v-271f3949=\\"\\"></label><!-- book author(s) -->
      <fieldset class=\\"add-form__fieldset\\" data-v-271f3949=\\"\\">
        <legend data-v-271f3949=\\"\\">Authors</legend>
        <div class=\\"add-form__author\\" data-v-271f3949=\\"\\">
          <!-- author's name --><label class=\\"add-form__item\\" data-v-271f3949=\\"\\"> Author's name* <input type=\\"text\\" data-test=\\"author-name-input\\" class=\\"add-form__input\\" data-v-271f3949=\\"\\"></label><label class=\\"add-form__item\\" data-v-271f3949=\\"\\"> Birth year* <input type=\\"text\\" data-test=\\"author-birth-input\\" class=\\"add-form__input\\" data-v-271f3949=\\"\\"></label><label class=\\"add-form__item\\" data-v-271f3949=\\"\\"> Death year <input type=\\"text\\" class=\\"add-form__input\\" data-v-271f3949=\\"\\"></label>
        </div><button data-test=\\"submit-button\\" class=\\"add-form__add-author\\" data-v-271f3949=\\"\\"> Add another author </button>
      </fieldset><!-- book subjects --><label class=\\"add-form__item\\" data-v-271f3949=\\"\\"> Subjects <input type=\\"text\\" placeholder=\\"Subjects, separated by commas\\" data-test=\\"subjects-input\\" class=\\"add-form__input\\" data-v-271f3949=\\"\\"></label><label class=\\"add-form__item\\" data-v-271f3949=\\"\\"> Bookshalves <input type=\\"text\\" placeholder=\\"Bookshalves, separated by commas\\" data-test=\\"bookshalves-input\\" class=\\"add-form__input\\" data-v-271f3949=\\"\\"></label><button class=\\"add-form__button\\" data-v-271f3949=\\"\\">Add</button>
    </form>
  </div>
</main>"
`;
